doc/design.md

```diff
@@ -0,0 +1,156 @@
+# domain_bridge Design
+
+This article documents the design of the domain_bridge.
+
+## Background
+
+A ROS 2 network can be partitioned by assigning each *node* (more specifically, a *context* associated with the node), a *domain ID*.
+A domain ID is simply an integer.
+Nodes with different domain IDs cannot directly communicate with each other.
+This is useful for avoiding interference and reducing network traffic, for example.
+For more information, check out [this article about DDS domains from RTI](https://community.rti.com/static/documentation/connext-dds/5.2.3/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_UsersManual/Content/UsersManual/Fundamentals_of_DDS_Domains_and_DomainPa.htm).
+
+However, there are applications that would like to be able to share *some* data between different ROS 2 domains.
+Therefore, it would be nice to have the ability to "bridge" a smaller set of ROS network traffic across domains, hence the *domain_bridge*.
+
+## Requirements
+
+The following are a set of requirements we would like our domain bridge to meet:
+
+1. Bridge ROS topics from one domain to another.
+
+This is pretty self-explanatory.
+Given a node *A* on domain *D<sub>A</sub>* and a node *B* on domain *D<sub>B</sub>*, where D<sub>A</sub> != D<sub>B</sub>,
+A and B should be able to communicate with the ROS pub/sub mechanism over one or more topics.
+
+2. Bridge ROS services from one domain to another.
+
+The same as (1), but for ROS services.
+
+3. Bridge ROS actions from one domain to another.
+
+The same as (1), but for ROS actions.
+
+4. No explicit dependencies on ROS interface types
+
+The bridge should work with any supported ROS type (e.g. .msg, .srv, .action, or .idl).
+Furthermore, the bridge should not have to declare any explicit dependencies on interfaces types (e.g. in code or package.xml).
+Rather, the bridge should dynamically load type support at runtime (and produce an error if support for the requested type does not exist).
+
+5. Accurately bridge Quality of Service settings
+
+All ROS communication entities (e.g. publishers and service clients) have [Quality of Service (QoS)](http://design.ros2.org/articles/qos.html) settings associated with them.
+The bridge should faithfully map the QoS settings of data coming from one domain into another domain.
+For example, a publisher with reliability policy set to "best effort" should continue to publish as "best effort" in the other domains when bridged.
```

Comment:
Good point, I meant to address that in the design doc but forgot. In order to handle any sort of edge cases, I think it's best to create a one-to-one mapping for each publisher being bridged. In your example, we would create two sus/pub pairs in the bridge. The first set would use "best effort" and the second set would use "reliable". This way we can hopefully match the exact QoS for each publisher being bridged. This will hopefully avoid any issues overflowing history due to insufficient history depth.

Leading me to another question: how should be handle history and depth policies? AFAIK, we can't query that info from the publisher endpoint, so maybe the best we can do is expose it as a config option in the bridge.

Reply:
> Leading me to another question: how should be handle history and depth policies? AFAIK, we can't query that info from the publisher endpoint, so maybe the best we can do is expose it as a config option in the bridge.

hmm.. nevermind, after some experimentation it seems we can query history and depth after all. I guess I was lead to believe it wasn't possible due to the [this API not returning them](https://github.com/ros2/rclpy/blob/3e030c09a3fbb1a5c4ac1f7d827c1deffff5d242/rclpy/rclpy/topic_endpoint_info.py#L160).

Reply:
I've added to the section in d7797d6

Reply:
> hmm.. nevermind, after some experimentation it seems we can query history and depth after all.

how? AFAIR that wasn't possible

Reply:
See https://github.com/ros2/rclpy/pull/686
I'm proposing adding history and depth to the string output of endpoint info, which shows up in `ros2 topic info --verbose`. It seems to work locally, but perhaps it's because of shared memory transport or something?

I tested by running a talker,

    ros2 run demo_nodes_cpp talker

and then using `ros2topic` (with the mentioned PR):

    ros2 topic info --verbose /chatter

Reply:
I don't know *how* it works, just reporting what I've observed.

Reply:
does it work with all rmw implementations?

AFAIR, that information is not always available (e.g. I'm almost sure it's not in fastrtps).

Reply:
I checked again, it looks like we can't get that information from fastrtps or connext.

Reply:
I've added a section to the design doc discussing this point: c420562